### PR TITLE
[SMTChecker] Support arithmetic compound assignment operators.

### LINKED ISSUE #6006: [SMTChecker] Compound assignment operators

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes #6006

### CODE REVIEW

File: libsolidity/formal/SMTChecker.cpp [hunk 1]
```diff
@@ -333,27 +333,54 @@ void SMTChecker::endVisit(VariableDeclarationStatement const& _varDecl)
...
 			"Assertion checker does not yet implement type " + _assignment.annotation().type->toString()
 		);
-	else if (Identifier const* identifier = dynamic_cast<Identifier const*>(&_assignment.leftHandSide()))
+	else if (auto identifier = dynamic_cast<Identifier const*>(&_assignment.leftHandSide()))
 	{
 		VariableDeclaration const& decl = dynamic_cast<VariableDeclaration const&>(*identifier->annotation().referencedDeclaration);
 		solAssert(knownVariable(decl), "");
-		assignment(decl, _assignment.rightHandSide(), _assignment.location());
-		defineExpr(_assignment, expr(_assignment.rightHandSide()));
+		smt::Expression rightHandSide = compoundToArithmetic.count(op) ?
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.cpp]:
```suggestion
		smt::Expression rightHandSide =
			compoundToArithmetic.count(op) ?
```

File: libsolidity/formal/SMTChecker.cpp [hunk 2]
```diff
@@ -333,27 +333,54 @@ void SMTChecker::endVisit(VariableDeclarationStatement const& _varDecl)
...
+			expr(_assignment.rightHandSide())
+		;
+		defineExpr(_assignment, rightHandSide);
+		assignment(decl, _assignment, _assignment.location());
 	}
 	else if (dynamic_cast<IndexAccess const*>(&_assignment.leftHandSide()))
 	{
-		arrayIndexAssignment(_assignment);
-		defineExpr(_assignment, expr(_assignment.rightHandSide()));
+		smt::Expression rightHandSide = compoundToArithmetic.count(op) ?
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.cpp]: Can this code be combined with the one above?

**Comment by chriseth** [comment]
If Solidity supports `a += b += c` - can you add a test for it?

**Comment by chriseth** [comment]
Actually please try both `(a += b) += c` and `a += (b += c)` :)

File: libsolidity/formal/SMTChecker.cpp (hunk 2 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTChecker.cpp]: I thought about that, the problem is that we need an `if identifier else if indexAccess` to know which parameters to use, then an `if compound` to know which operation to call, and then another `if identifier else if indexAccess` to write correctly...

**Comment by leonardoalt** [comment]
Solidity accepts `a += b += c`, `a += (b += c)` but doesn't `(a += b) += c`, `b = (a += b) += c`, and `a += (a += b) += c`.

File: libsolidity/formal/SMTChecker.cpp (hunk 2 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTChecker.cpp]: Attempted to make it look nicer, updated.

**Comment by leonardoalt** [comment]
Updated with new test

**Review by chriseth — APPROVED** [review]

**Pull request merged**